### PR TITLE
Pathplanner gcs fixes part 2

### PR DESCRIPTION
**Pull request by elafargue**
This adds more consistency checks in the path planner import/export functions, as well as removes the "Waypoint Import/Export" from the Files... menu, which was duplicating the functionality of the Path Planner module in an incompatible way.

### CODE REVIEW

**Comment by elafargue** [comment]
This fixes #1612

File: ground/gcs/src/plugins/gcscontrolwidget/gcscontrolgadget.cpp [hunk 1]
```diff
@@ -125,7 +125,7 @@ GCSControl* GCSControlGadget::getGcsControl() {
 void GCSControlGadget::manualControlCommandUpdated(UAVObject * obj) {
 
     // Not sending then show updates from transmitter
-    if (!enableSending)
+    if (enableSending)
```
Comment (peabody124) [inline ground/gcs/src/plugins/gcscontrolwidget/gcscontrolgadget.cpp]: You mind trying to rebase and force push your branch? Looks like we are getting some changes from a previous commit in this diff.

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp [hunk 2]
```diff
@@ -520,7 +515,9 @@ void FlightDataModel::readFromFile(QString fileName)
     // For now, reset home location to the location in the file. Later we can see if we would
     // rather ask the user whether to remap the flight plan to the current home location or use
     // the one in the map
-    setHomeLocation(HomeLLA);
+    if (!setHomeLocation(HomeLLA)) {
+        showErrorDialog("Home location error", "Home location coordinates invalid");
+    }
```
Comment (mlyle) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: Does it make sense to keep going here?  I can see arguments either way.

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp [hunk 3]
```diff
@@ -534,28 +531,45 @@ void FlightDataModel::readFromFile(QString fileName)
...
                         data->locked = field.attribute("value").toInt();
+                        params++;
+                    }
                 }
                 fieldNode=fieldNode.nextSibling();
             }
+
+            // We can't really check everything in the file is fully consistent, but
+            // at least we can make sure we have the right number of fields
+            if (params < 8) {
```
Comment (mlyle) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: Will this catch if a field is duplicated?  Might make sense to have ceilings on north/east/down/velocity, too.

**Comment by mlyle** [comment]
Looks good to me.  I assume you've tested it?

File: ground/gcs/src/plugins/opmap/opmapgadget.cpp [hunk 4]
```diff
@@ -73,7 +73,7 @@ void OPMapGadget::loadConfiguration(IUAVGadgetConfiguration *config)
     m_widget->SetUavPic(m_config->uavSymbol());
     m_widget->setZoom(m_config->zoom());
     m_widget->setPosition(QPointF(m_config->longitude(), m_config->latitude()));
-    m_widget->setHomePosition(QPointF(m_config->longitude(), m_config->latitude()));
+    // m_widget->setHomePosition(QPointF(m_config->longitude(), m_config->latitude()));
```
Comment (peabody124) [inline ground/gcs/src/plugins/opmap/opmapgadget.cpp]: delete commented code

File: ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp [hunk 5]
```diff
@@ -265,6 +265,18 @@ OPMapGadgetWidget::OPMapGadgetWidget(QWidget *parent) : QWidget(parent)
         }
     }
 
+    // Update the home location if it is already set:
+    if (obum) {
```
Comment (peabody124) [inline ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp]: not your variable, but what is obum? rename?

File: ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp [hunk 6]
```diff
@@ -265,6 +265,18 @@ OPMapGadgetWidget::OPMapGadgetWidget(QWidget *parent) : QWidget(parent)
...
+        bool set;
+        double LLA[3];
+        if (obum->getHomeLocation(set, LLA) == 0) {
+            if (LLA[0] != 0 && LLA[1] != 0 && LLA[2] != 0)
```
Comment (peabody124) [inline ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp]: Check the set variable?

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp [hunk 7]
```diff
@@ -520,7 +515,9 @@ void FlightDataModel::readFromFile(QString fileName)
     // For now, reset home location to the location in the file. Later we can see if we would
```
Comment (peabody124) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: This comment appears out of date.

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp (hunk 3 above)
Comment (peabody124) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: seems like a reasonable idea, but could be a future enhancement.

File: ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp (hunk 6 above)
Comment (elafargue) [inline ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp]: We don't care about it in this particular context (the set flag is only used to check that we have set the home location in the UAV, unless I misunderstand its use ? )

File: ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.cpp [hunk 8]
```diff
@@ -404,7 +265,6 @@ QString UAVSettingsImportExportFactory::createXMLDocument(const enum storedData
     // create settings and/or data elements
     QDomElement settings = doc.createElement("settings");
     QDomElement data = doc.createElement("data");
-    QDomElement waypoints = doc.createElement("waypoints");
```
Comment (peabody124) [inline ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.cpp]: Ok. So just to be clear we had _three_ waypoint representations? This was adding it to a standard settings export too, and that is what you are removing?

File: ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.cpp [hunk 9]
```diff
@@ -586,11 +386,11 @@ void UAVSettingsImportExportFactory::exportUAVSettings()
 void UAVSettingsImportExportFactory::exportUAVData()
 {
     if (QMessageBox::question(0, tr("Are you sure?"),
-        tr("This option is only useful for passing your current "
-           "system data to the technical support staff. "
-           "Do you really want to export?"),
-           QMessageBox::Ok | QMessageBox::Cancel,
-           QMessageBox::Ok) != QMessageBox::Ok) {
+                              tr("This option is only useful for passing your current "
+                                 "system data to the technical support staff. "
```
Comment (peabody124) [inline ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.cpp]: Awww :( we are staff? Maybe developers? :-p

File: ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.cpp (hunk 8 above)
Comment (elafargue) [inline ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.cpp]: That's correct, we had three, all slightly overlapping but not entirely consistent. I don't see the point of including a fixed flight plan in a UAV Config, this sounds very dangerous.

File: ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.cpp (hunk 9 above)
Comment (elafargue) [inline ground/gcs/src/plugins/uavsettingsimportexport/uavsettingsimportexportfactory.cpp]: Maybe we should get cool "Staff" T-Shirts with the TL logo ?

File: ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp [hunk 10]
```diff
@@ -98,8 +98,8 @@ OPMapGadgetWidget::OPMapGadgetWidget(QWidget *parent) : QWidget(parent)
     m_mouse_waypoint = NULL;
 
 	pm = NULL;
-	obm = NULL;
-	obum = NULL;
+    uavo_mgr = NULL;
```
Comment (peabody124) [inline ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp]: After this PR is merged, could you make a new PR that reinvents this whole file to spaces? It is currently mixed and makes it look like you messed up indentation in PR (I checked, you are using spaces appropriately)

File: ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp [hunk 11]
```diff
@@ -265,6 +265,18 @@ OPMapGadgetWidget::OPMapGadgetWidget(QWidget *parent) : QWidget(parent)
         }
     }
 
+    // Update the home location if it is already set:
+    if (uavo_util_mgr) {
+        bool set;
+        double LLA[3];
+        if (uavo_util_mgr->getHomeLocation(set, LLA) == 0) {
+            if (LLA[0] != 0 && LLA[1] != 0 && LLA[2] != 0)
```
Comment (peabody124) [inline ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp]: It looks like you still aren't checking the "set" variable here. Shouldn't you?

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp [hunk 12]
```diff
@@ -476,12 +476,7 @@ void FlightDataModel::readFromFile(QString fileName)
     QByteArray array=file.readAll();
     QString error;
     if (!doc.setContent(array,&error)) {
-        QMessageBox msgBox;
-        msgBox.setText(tr("File Parsing Failed."));
-        msgBox.setInformativeText(QString(tr("This file is not a correct XML file:%0")).arg(error));
-        msgBox.setStandardButtons(QMessageBox::Ok);
-        msgBox.exec();
-        return;
+        showErrorDialog("File Parsing Failure", "This file is not a correct XML file");
```
Comment (peabody124) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: tr()

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp [hunk 13]
```diff
@@ -517,10 +512,12 @@ void FlightDataModel::readFromFile(QString fileName)
         homelocField = homelocField.nextSibling();
     }
 
-    // For now, reset home location to the location in the file. Later we can see if we would
-    // rather ask the user whether to remap the flight plan to the current home location or use
-    // the one in the map
-    setHomeLocation(HomeLLA);
+    // For now, reset home location to the location in the file.
```
Comment (peabody124) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: These comments are still out of date, yes?

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp [hunk 14]
```diff
@@ -517,10 +512,12 @@ void FlightDataModel::readFromFile(QString fileName)
...
+    // In a future revision, we should consider asking the user whether to remap the flight plan
+    // to the current home location or use the one in the flight plan
+    if (!setHomeLocation(HomeLLA)) {
+        showErrorDialog("Home location error", "Home location coordinates invalid");
```
Comment (peabody124) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: tr()

File: ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp (hunk 11 above)
Comment (elafargue) [inline ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp]: I don't think I should check it there, unless there is something I don't
understand about the use of that flag?

**Comment by peabody124** [comment]
Just a few more small things :)

File: ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp (hunk 11 above)
Comment (peabody124) [inline ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp]: Well if Set is false, the home location is functionally invalid so we
should treat it like it is not defined.

Comment (elafargue) [inline ground/gcs/src/plugins/opmap/opmapgadgetwidget.cpp]: I thought it always turned back to false after it is set? Ok, probably my mistake, I'll update.

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp (hunk 13 above)
Comment (elafargue) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: They are not: we do reset the home location. The comment mentions that in a future release, we should let the user select whether to update the home location, or to remap the flight plan to the currently stored home location on the uav

File: ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp (hunk 14 above)
Comment (elafargue) [inline ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp]: tr() is done in showErrorDialog

**Comment by mlyle** [comment]
I tried this, it seemed to mostly work.  I got a couple crashes but nothing reproducible.

My main concern is this:  if I delete waypoints, download them to the flight controller, and then download them from the flight controller... they come back.  So I think deletion is not really working.

**Comment by elafargue** [comment]
@mlyle : agree, transfer of the waypoints to/from the flight controller is problematic, but is another issue, this PR does not try to fix it. I'm torn between making a huge PR that attempts to fix everything and will never go anywhere, or tackle each problem one after the other.

**Comment by mlyle** [comment]
OK, if that's not in scope for this one I'm cool with this landing.

**Pull request merged**